iceoryx_hoofs/test/moduletests/test_polymorphic_handler.cpp

```diff
@@ -0,0 +1,249 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/design_pattern/polymorphic_handler.hpp"
+#include "iceoryx_hoofs/design_pattern/static_lifetime_guard.hpp"
+
+#include "test.hpp"
+#include <iostream>
+
+namespace
+{
+using namespace ::testing;
+
+struct Interface : public iox::design_pattern::Activatable
+{
+    virtual ~Interface() = default;
+
+    virtual uint32_t id() = 0;
+
+    void reset()
+    {
+        value = 0;
+    }
+
+    uint32_t value{0};
+};
+
+constexpr uint32_t DEFAULT_ID = 73;
+constexpr uint32_t ALTERNATE_ID = 21;
+
+struct Default : public Interface
+{
+    uint32_t id() override
+    {
+        return DEFAULT_ID;
+    }
+};
+
+struct Alternate : public Interface
+{
+    uint32_t id() override
+    {
+        return ALTERNATE_ID;
+    }
+};
+
+template <typename T>
+using Guard = iox::design_pattern::StaticLifetimeGuard<T>;
+
+// should the handler instances be accessed, they will live at least as
+// long as the guard objects
+Guard<Default> defaultGuard;
+Guard<Alternate> alternateGuard;
+
+// will live at least as long as the corresponding guards
+Default& defaultHandler = Guard<Default>::instance();
+Alternate& alternateHandler = Guard<Alternate>::instance();
+
+struct Hooks
+{
+    // to check whether the arguments are used correctly
+    static void onSetAfterFinalize(Interface& currentInstance, Interface& newInstance)
+    {
+        currentInstance.value = currentInstance.id();
+        newInstance.value = newInstance.id();
+    }
+};
+
+using Handler = iox::design_pattern::PolymorphicHandler<Interface, Default, Hooks>;
+
+class PolymorphicHandler_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+        Handler::reset();
+    }
+
+    void TearDown() override
+    {
+    }
+};
+
+TEST_F(PolymorphicHandler_test, handlerIsInitializedWithDefault)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "41bb4a5e-a916-4a6d-80c4-fed3a3d8d78b");
+    auto& handler = Handler::get();
+
+    EXPECT_EQ(handler.id(), DEFAULT_ID);
+}
+
+TEST_F(PolymorphicHandler_test, settingAlternateWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8b2f0cfe-f13c-4fa0-aa93-5ddd4f0904d1");
+    Handler::set(alternateGuard);
+    auto& handler = Handler::get();
+
+    EXPECT_EQ(handler.id(), ALTERNATE_ID);
+}
+
+TEST_F(PolymorphicHandler_test, alternatePointsToExternalMemory)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "85ce0e51-a1fe-490c-9012-7d539512ed38");
+    Handler::set(alternateGuard);
+
+    auto& handler = Handler::get();
+    auto* ptr = static_cast<Interface*>(&alternateHandler);
+
+    EXPECT_EQ(&handler, ptr);
+}
+
+TEST_F(PolymorphicHandler_test, explicitlySettingToDefaultWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "32e4d808-c848-4bf9-b878-e163ca825539");
+    Handler::set(alternateGuard);
+    Handler::set(defaultGuard);
```

Comment:
Can we add a test where we check whether the return value of `set` equals the previous instance and if the handler is deactivated?

Reply:
Yes. Of course, previous is only controllable if it is not used concurrently. Concurrent correctness is the weakness in general. 

However, as presented sometime ago, sequential correctness is a necessary condition for concurrent correctness and here we gather evidence for sequential correctness (cannot prove correctness via tests anyway).

Reply:
I added this to tests and also some specific tests that deal with setting of the same handler that is set or resetting to default from default.